includes/Product/Shopify_Query.php

```diff
@@ -0,0 +1,239 @@
+		$this->settings = $settings;
+	}
+
+	/**
+	 * Get products by search term.
+	 *
+	 * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+	 * @SuppressWarnings(PHPMD.NPathComplexity)
+	 *
+	 * @since 1.20.0
+	 *
+	 * @param string $search_term Search term.
+	 * @return void|WP_Error
+	 */
+	public function do_search( string $search_term ) {
+
+		/**
+		 * Filters the shopify products data TTL value.
+		 *
+		 * @since 1.20.0
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_shopify_data_cache_ttl', DAY_IN_SECONDS ); // TODO: lower default TTL.
+		$cache_key = ! empty( $search_term ) ? 'web_stories_shopify_data_' . md5( $search_term ) : 'web_stories_shopify_data_default';
+
+		$data = get_transient( $cache_key );
+
+		if ( \is_string( $data ) && ! empty( $data ) ) {
+			/**
+			 * Decoded cached products data.
+			 *
+			 * @var array|null $products
+			 */
+			$products = json_decode( $data, true );
+
+			if ( \is_array( $products ) ) {
+				$results = [];
+				foreach ( $products as $product ) {
+					$results[] = new Product( $product );
+				}
+				$this->set_results( $results );
+				return;
+			}
+		}
+
+		$results     = [];
+		$results_raw = [];
+
+		/**
+		 * Shopify host.
+		 *
+		 * @var string $host
+		 */
+		$host = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_HOST );
+
+		/**
+		 * Shopify access token.
+		 *
+		 * @var string $access_token
+		 */
+		$access_token = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_ACCESS_TOKEN );
+
+		// TODO(#11154): Maybe move to a class constant.
+		$api_version = '2022-01';
+
+		// TODO(#11154): Proper sanitization.
+		$url = sprintf(
+			'https://%1$s/api/%2$s/graphql.json',
+			$host,
+			$api_version
+		);
+
+		$search_string = empty( $search_term ) ? '*' : '*' . $search_term . '*';
+
+		// TODO(#11154): Support different sortKeys.
+		// Maybe use "available_for_sale:true AND " query to only show items in stock.
+		$query = <<<QUERY
+{
+  products(first: 100, sortKey: CREATED_AT, query: "title:$search_string") {
+    edges {
+      node {
+        id
+        handle
+        title
+        vendor
+        description
+        priceRange {
+          minVariantPrice {
+            amount
+            currencyCode
+          }
+        }
+        onlineStoreUrl
+        images(first: 10) {
+          edges {
+            node {
+              url
+              altText
+            }
+          }
+        }
+      }
+    }
+  }
+}
+QUERY;
+
+		$response = wp_remote_post(
+			$url,
+			[
+				'headers' => [
+					'Content-Type'                      => 'application/graphql',
+					'X-Shopify-Storefront-Access-Token' => $access_token,
+				],
+				'body'    => $query,
+			]
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			return new WP_Error( 'rest_unknown', __( 'Error fetching products', 'web-stories' ), [ 'status' => 404 ] );
```

Comment:
Can we use the real response status here? With the current code on `main` the 404 status code led me to believe I had a different issue when I was troubleshooting something.

Reply:
Again, just a copy and paste from elsewhere.

Reply:
Improving error handling is left to be done in #11268